### PR TITLE
feat: V4 POST policy

### PR DESCRIPTION
**Pull request by JesseLovelace**
Adds V4 post policy. Will add documentation once implementation is approved 

Fixes: #280 #88

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 1]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+     *     Accounts</a>
+     */
+    public static V4PostPolicyOption signWith(ServiceAccountSigner signer) {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.SERVICE_ACCOUNT_CRED, signer);
+    }
+
+    /**
+     * Use a virtual hosted-style hostname, which adds the bucket into the host portion of the URI
+     * rather than the path, e.g. 'https://mybucket.storage.googleapis.com/...'. The bucket name
+     * will be obtained from the resource passed in.
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: will be --> is
per Google Tech Writing Guidelines
https://developers.google.com/style/tense

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 2]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+    V4PostPolicyOption.Option getOption() {
+      return option;
+    }
+
+    Object getValue() {
+      return value;
+    }
+
+    /**
+     * Provides a service account signer to sign the policy. If not provided an attempt will be made
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: will be --> is

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 3]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints">Request Endpoints</a>
+     */
+    public static V4PostPolicyOption withVirtualHostedStyle() {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.VIRTUAL_HOSTED_STYLE, "");
+    }
+
+    /**
+     * Generate a path-style URL, which places the bucket name in the path portion of the URL
+     * instead of in the hostname, e.g 'https://storage.googleapis.com/mybucket/...'. Note that this
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: delete "Note that"
https://developers.google.com/style/tone

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 4]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+     * will be obtained from the resource passed in.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints">Request Endpoints</a>
+     */
+    public static V4PostPolicyOption withVirtualHostedStyle() {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.VIRTUAL_HOSTED_STYLE, "");
+    }
+
+    /**
+     * Generate a path-style URL, which places the bucket name in the path portion of the URL
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Generate --> Generates
https://developers.google.com/style/reference-verbs

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 5]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints">Request Endpoints</a>
+     */
+    public static V4PostPolicyOption withPathStyle() {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.PATH_STYLE, "");
+    }
+
+    /**
+     * Use a bucket-bound hostname, which replaces the storage.googleapis.com host with the name of
+     * a CNAME bucket, e.g. a bucket named 'gcs-subdomain.my.domain.tld', or a Google Cloud Load
+     * Balancer which routes to a bucket you own, e.g. 'my-load-balancer-domain.tld'. Note that this
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: delete "Note that"

Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: No Latin; e.g. --> for example
https://developers.google.com/style/abbreviations

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 6]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+     *     GCLB Redirects</a>
+     */
+    public static V4PostPolicyOption withBucketBoundHostname(String bucketBoundHostname) {
+      return withBucketBoundHostname(bucketBoundHostname, UriScheme.HTTP);
+    }
+
+    /**
+     * Use a bucket-bound hostname, which replaces the storage.googleapis.com host with the name of
+     * a CNAME bucket, e.g. a bucket named 'gcs-subdomain.my.domain.tld', or a Google Cloud Load
+     * Balancer which routes to a bucket you own, e.g. 'my-load-balancer-domain.tld'. Note that this
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: No Latin; e.g. --> for example
https://developers.google.com/style/abbreviations

Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: delete "Note that"

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 7]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+  class V4PostPolicy {
+    private String url;
+    private Map<String, String> fields;
+
+    private V4PostPolicy(String url, Map<String, String> fields) {
+      this.url = url;
+      this.fields = fields;
+    }
+
+    public static V4PostPolicy of(String url, Map<String, String> fields) {
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: These methods don't need to be public since the class isn't

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 8]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+     */
+    public static V4PostPolicyOption withBucketBoundHostname(
+        String bucketBoundHostname, UriScheme uriScheme) {
+      return new V4PostPolicyOption(
+          V4PostPolicyOption.Option.BUCKET_BOUND_HOST_NAME,
+          uriScheme.getScheme() + "://" + bucketBoundHostname);
+    }
+  }
+
+  class V4PostPolicy {
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: This class looks like it could be static, and per effective Java if it can be it should. Outer class would be even clearer.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 9]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+    public String getUrl() {
+      return url;
+    }
+
+    public Map<String, String> getFields() {
+      return fields;
+    }
+  }
+
+  class V4PostFields {
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: same comments as above; make static and outer if possible, no public

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 10]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+      this.cacheControl = builder.cacheControl;
+      this.contentDisposition = builder.contentDisposition;
+      this.contentEncoding = builder.contentEncoding;
+      this.contentLength = builder.contentLength;
+      this.contentType = builder.contentType;
+      this.expires = builder.expires;
+      this.successActionRedirect = builder.successActionRedirect;
+      this.successActionStatus = builder.successActionStatus;
+
+      this.fieldsMap = new HashMap<>();
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: There's a constructor and builder and they're both setting the fieldsMap to a new mao. This is at least hard to follow and might indeed be a big. Making fieldsMap final and initializing it on the field declaration instead of in the constructor would eliminate this possibility.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 11]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+      }
+
+      public Builder setSuccessActionStatus(int successActionStatus) {
+        this.successActionStatus = "" + successActionStatus;
+        return this;
+      }
+    }
+  }
+
+  class V4PostConditions {
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: same as above. This is too many inner classes, and the outer class is over 1000 lines. For sheer size, try breaking this up into independent, outer package protected classes.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 12]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
...
+    private V4PostPolicyDocument(String expiration, V4PostConditions conditions) {
+      this.expiration = expiration;
+      this.conditions = conditions;
+    }
+
+    public static V4PostPolicyDocument of(String expiration, V4PostConditions conditions) {
+      return new V4PostPolicyDocument(expiration, conditions);
+    }
+
+    public String toJsonObject() {
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: I'm not sure about this one but consider toJson, not to JsonObject in keeping with return type

File: google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java [hunk 13]
```diff
@@ -733,6 +736,124 @@ public URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOptio
...
+            && optionMap.containsKey(SignUrlOption.Option.PATH_STYLE)
+            && optionMap.containsKey(SignUrlOption.Option.BUCKET_BOUND_HOST_NAME)),
+        "Only one of VIRTUAL_HOSTED_STYLE, PATH_STYLE, or BUCKET_BOUND_HOST_NAME SignUrlOptions can be"
+            + " specified.");
+
+    String bucketName = slashlessBucketNameFromBlobInfo(blobInfo);
+
+    boolean usePathStyle = shouldUsePathStyleForSignedUrl(optionMap);
+
+    String url =
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java]: The ternary operator is hard to read when broken across multiple lines. Consider if else.

File: google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java [hunk 14]
```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2019 Google LLC
```
Comment (elharo) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java]: 2020

File: google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import static com.google.cloud.storage.Storage.V4ConditionType.MATCHES;
```
Comment (elharo) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java]: Use non-static imports here as these are not well-known names.

File: google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+    }
+  }
+
+  @Rule public TestName testName = new TestName();
+
+  private final PostPolicyV4Test testData;
+  private final ServiceAccountCredentials serviceAccountCredentials;
+
+  /**
+   * @param testData The serialized test data representing the test case.
```
Comment (elharo) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java]: nit: no cap and no period since it's not a complete sentence

File: google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+            blob, v4Fields, builder.build(), testData.getPolicyInput().getExpiration(), style);
+    assertEquals(testData.getPolicyOutput().getFieldsMap(), policy.getFields());
+    assertEquals(
+        testData.getPolicyOutput().getExpectedDecodedPolicy(),
+        new String(BaseEncoding.base64().decode(policy.getFields().get("policy"))));
+    assertEquals(testData.getPolicyOutput().getUrl(), policy.getUrl());
+  }
+
+  /**
+   * Load all of the tests and return a {@code Collection<Object[]>} representing the set of tests.
```
Comment (elharo) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java]: Load --> Loads

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java (hunk 11 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: +1 could you split this up into separate files for maintainability as well?

File: google-cloud-storage/src/test/java/com/google/cloud/storage/V4SigningTest.java [hunk 18]
```diff
@@ -104,14 +105,14 @@ public void test() {
 
     SignUrlOption style = SignUrlOption.withPathStyle();
 
-    if (testData.getUrlStyle().equals(SigningV4Test.UrlStyle.VIRTUAL_HOSTED_STYLE)) {
+    if (testData.getUrlStyle().equals(UrlStyle.VIRTUAL_HOSTED_STYLE)) {
```
Comment (frankyn) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/V4SigningTest.java]: can you split out this into a separte PR if possible?

Comment (frankyn) [reply google-cloud-storage/src/test/java/com/google/cloud/storage/V4SigningTest.java]: For this specifically, I mean to split out conformance test changes into a separate PR.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 19]
```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
     }
   }
 
+  class V4PostPolicyOption implements Serializable {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Can you split all V4PostPolicy helper classes into a separate file, except what's needed for `generateV4PresignedPostPolicy` ease of use such as `V4PostPolicyOption`?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java [hunk 20]
```diff
@@ -733,6 +736,124 @@ public URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOptio
...
+    return generateV4PresignedPostPolicy(
+        blobInfo, fields, V4PostConditions.newBuilder().build(), duration, options);
+  }
+
+  public V4PostPolicy generateV4PresignedPostPolicy(
+      BlobInfo blobInfo, long duration, V4PostPolicyOption... options) {
+    return generateV4PresignedPostPolicy(
+        blobInfo, V4PostFields.newBuilder().build(), duration, options);
+  }
+
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java]: Put V4 at the end of the name please.

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java]: Please do this for other helpers as well that use V4 for POST Policy. From meeting.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 21]
```diff
@@ -2532,6 +2978,13 @@ Blob create(
    */
   URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOption... options);
 
+  V4PostPolicy generateV4PresignedPostPolicy(
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Using consistent naming across languages, please use: `generateSignedPostPolicyV4`.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 22]
```diff
@@ -2532,6 +2639,46 @@ Blob create(
...
+   *
+   * }</pre>
+   *
+   * @param blobInfo the blob uploaded in the form
+   * @param fields the fields specified in the form
+   * @param conditions which conditions every upload must satisfy
+   * @param duration how long until the form expires, in milliseconds
+   * @param options optional post policy options
+   */
+  PostPolicyV4 generatePresignedPostPolicyV4(
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2674]: use `generateSignedPostPolicyV4`

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 23]
```diff
@@ -2532,6 +2639,46 @@ Blob create(
...
+      BlobInfo blobInfo,
+      PostFieldsV4 fields,
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2676]: fields and conditions should be optional. I want to say these should be under options.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 24]
```diff
@@ -0,0 +1,378 @@
...
+        String element;
+        String value;
+
+        private ConditionV4(ConditionV4Type type, String element, String value) {
+            this.type = type;
+            this.element = element;
+            this.value = value;
+        }
+
+        public static ConditionV4 newV4Condition(ConditionV4Type type, String element, String value) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:361]: This doesn't translate correctly for ContentRangeLengths because element and value are start  and end respectively.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 25]
```diff
@@ -0,0 +1,378 @@
...
+            return new ConditionV4(type, element, value);
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:362]: Given PostConditionV4 requires users to use a related method. Should this class have a private static method helper as well? Also not sure how this is different from `private ConditionV4()`?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 26]
```diff
@@ -0,0 +1,374 @@
...
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Presigned V4 post policy.
+ *
+ * @see <a href="https://cloud.google.com/storage/docs/xml-api/post-object" POST Object</a>
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:17]: Missing `>`

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 27]
```diff
@@ -0,0 +1,374 @@
...
+    }
+
+    public Map<String, String> getFields() {
+        return fields;
+    }
+
+    /**
+     * Class representing which fields to specify in a V4 POST request.
+     *
+     * @see <a href=https://cloud.google.com/storage/docs/xml-api/post-object#form_fields> POST Object Form fields</a>
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:43]: No double quote `"` around the URL.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 28]
```diff
@@ -0,0 +1,374 @@
...
+                return this;
+            }
+
+            public Builder addContentLengthRangeCondition(int min, int max) {
+                conditions.add(
+                        new ConditionV4(ConditionV4Type.CONTENT_LENGTH_RANGE, "" + min, "" + max));
+                return this;
+            }
+
+            public Builder addCustomCondition(ConditionV4Type type, String field, String value) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:261]: Suggest removing this until we get a feature request for this method.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 29]
```diff
@@ -2532,6 +2639,56 @@ Blob create(
...
+   * PostPolicyV4Option#signWith(ServiceAccountSigner)} option.
+   *
+   * <p>Example of generating a presigned post policy which has the condition that only jpeg images
+   * can be uploaded, and applies the public read acl to each image uploaded:
+   *
+   * <pre>{@code
+   * PostFieldsV4 fields = PostFieldsV4.newBuilder().setAcl("public-read").build();
+   * PostConditionsV4 conditions = PostConditionsV4.newBuilder().addContentTypeCondition(ConditionV4Type.MATCHES, "image/jpeg").build();
+   *
+   * PostPolicyV4 policy = storage.generatePresignedPostPolicyV4(
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2662]: uses old, instead of generateSignedPostPolicyV4.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 30]
```diff
@@ -2532,6 +2639,56 @@ Blob create(
...
+   *     BlobInfo.newBuilder("my-bucket", "my-object").build(),
+   *     fields, conditions, 7, TimeUnit.DAYS);
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2664]: Could you add an example of how to send a POST request with the policy result?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 31]
```diff
@@ -2532,6 +2639,56 @@ Blob create(
...
+   *
+   * }</pre>
+   *
+   * @param blobInfo the blob uploaded in the form
+   * @param fields the fields specified in the form
+   * @param conditions which conditions every upload must satisfy
+   * @param duration how long until the form expires, in milliseconds
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2671]: `unit` is missing here and `duration` is out of order.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java (hunk 13 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java:779]: Elliotte's comment here makes sense. Probably want to use an if-else in this case.

File: google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java (hunk 14 above)
Comment (frankyn) [reply google-cloud-storage/src/test/java/com/google/cloud/storage/V4PostPolicyTest.java:2]: Should be 2020.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 32]
```diff
@@ -2532,6 +2639,56 @@ Blob create(
...
+      PostConditionsV4 conditions,
+      PostPolicyV4Option... options);
+
+  PostPolicyV4 generateSignedPostPolicyV4(
+          BlobInfo blobInfo, long duration, TimeUnit unit, PostFieldsV4 fields, PostPolicyV4Option... options);
+
+  PostPolicyV4 generateSignedPostPolicyV4(
+          BlobInfo blobInfo, long duration, TimeUnit unit, PostConditionsV4 conditions, PostPolicyV4Option... options);
+
+  PostPolicyV4 generateSignedPostPolicyV4(
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2688]: Add integration tests.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 33]
```diff
@@ -0,0 +1,374 @@
...
+
+            public Builder setSuccessActionRedirect(String successActionRedirect) {
+                this.successActionRedirect = successActionRedirect;
+                return this;
+            }
+
+            public Builder setSuccessActionStatus(int successActionStatus) {
+                this.successActionStatus = "" + successActionStatus;
+                return this;
+            }
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:138]: Custom field setter should be available here.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 34]
```diff
@@ -0,0 +1,374 @@
...
+                        new ConditionV4(type, "success_action_redirect", successActionRedirectUrl));
+                return this;
+            }
+
+            public Builder addSuccessActionStatusCondition(ConditionV4Type type, int status) {
+                conditions.add(new ConditionV4(type, "success_action_status", "" + status));
+                return this;
+            }
+
+            public Builder addCustomMetadataCondition(ConditionV4Type type, String field, String value) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:250]: this should be in fields instead.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java (hunk 28 above)
Comment (JesseLovelace) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:261]: We actually use it in the generate method, but I can change it to package private

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:261]: package private sgtm, thanks for clarifying!

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 35]
```diff
@@ -2532,6 +2634,88 @@ Blob create(
...
+   * PostPolicyV4 policy = storage.generateSignedPostPolicyV4(
+   *     BlobInfo.newBuilder("my-bucket", "my-object").build(),
+   *     7, TimeUnit.DAYS, fields, conditions);
+   *
+   * HttpClient client = HttpClientBuilder.create().build();
+   * HttpPost request = new HttpPost(policy.getUrl());
+   * for(Map.Entry<String, String> entry : policy.getFields().entrySet()) {
+   *     request.addHeader(entry.getKey(), entry.getValue());
+   * }
+   * client.execute(request);
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2667]: This example is missing the file to upload*

File: google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java [hunk 36]
```diff
@@ -67,6 +67,8 @@
 import com.google.cloud.storage.CopyWriter;
 import com.google.cloud.storage.HmacKey;
 import com.google.cloud.storage.HttpMethod;
+import com.google.cloud.storage.PostPolicyV4;
+import com.google.cloud.storage.PostPolicyV4.*;
```
Comment (frankyn) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java:71]: Explicitly import the needed classes and not use `*`.

File: google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java [hunk 37]
```diff
@@ -3208,4 +3210,72 @@ public void testBucketLogging() throws ExecutionException, InterruptedException
       RemoteStorageHelper.forceDelete(storage, loggingBucket, 5, TimeUnit.SECONDS);
     }
   }
+
+  @Test
+  public void testSignedPostPolicyV4() {
```
Comment (frankyn) [inline google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java:3215]: This is more of a unit test so I would say it's good to keep. I'd like to see a request similar to what you added into your documentation above:

```
PostFieldsV4 fields = PostFieldsV4.newBuilder().setAcl("public-read").build();
PostConditionsV4 conditions = PostConditionsV4.newBuilder().addContentTypeCondition(ConditionV4Type.MATCHES, "image/jpeg").build();

PostPolicyV4 policy = storage.generateSignedPostPolicyV4(
 BlobInfo.newBuilder("my-bucket", "my-object").build(),
 7, TimeUnit.DAYS, fields, conditions);

HttpClient client = HttpClientBuilder.create().build();
HttpPost request = new HttpPost(policy.getUrl());
for(Map.Entry<String, String> entry : policy.getFields().entrySet()) {
     request.addHeader(entry.getKey(), entry.getValue());
}
// Missing data to upload in request
client.execute(request);
```

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 38]
```diff
@@ -2532,6 +2634,88 @@ Blob create(
    */
   URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOption... options);
 
+  /**
+   * Generates a URL and a map of fields that can be specified in a form to submit a POST request.
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:2638]: `in an HTML form`
You can link to relevant documentation here: https://cloud.google.com/storage/docs/xml-api/post-object#usage_and_examples

File: google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java [hunk 39]
```diff
@@ -0,0 +1,389 @@
...
+      public PostConditionsV4 build() {
+        return new PostConditionsV4(this);
+      }
+
+      public Builder addAclCondition(ConditionV4Type type, String acl) {
+        conditions.add(new ConditionV4(type, "acl", acl));
+        return this;
+      }
+
+      public Builder addBucketCondition(ConditionV4Type type, String bucket) {
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/PostPolicyV4.java:197]: Please remove this condition.

File: pom.xml [hunk 40]
```diff
@@ -86,6 +86,11 @@
         <type>pom</type>
         <scope>import</scope>
       </dependency>
+      <dependency>
+        <groupId>com.google.code.gson</groupId>
+        <artifactId>gson</artifactId>
+        <version>2.8.6</version>
```
Comment (frankyn) [inline pom.xml:92]: what's the scope for gson artifact?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 41]
```diff
@@ -56,7 +57,6 @@
  *
  * @see <a href="https://cloud.google.com/storage/docs">Google Cloud Storage</a>
  */
-@InternalExtensionOnly
 public interface Storage extends Service<StorageOptions> {
```
Comment (dmitry-fa) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:60]: @JesseLovelace , @frankyn 

InternalExtensionOnly is necessary, otherwise this change will break API.

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java:60]: Definitely missed this during review, thank you @dmitry-fa!

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java [hunk 42]
```diff
@@ -1989,73 +2089,60 @@ Blob create(
...
+   *     .build());
    * }</pre>
    *
-   * @param blobInfo information to update
    * @return the updated blob
    * @throws StorageException upon failure
-   * @see <a
-   *     href="https://cloud.google.com/storage/docs/json_api/v1/objects/update">https://cloud.google.com/storage/docs/json_api/v1/objects/update</a>
    */
   Blob update(BlobInfo blobInfo);
```
[the remark this answers is not in the document]
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java]: Addressed issues. Thanks @dmitry-fa
